### PR TITLE
Add Traveling Salesperson Algorithm

### PR DESCRIPTION
**Pull request by yjzhang111**
Add TSP Algorithm and testing files.

### CODE REVIEW

**Review by dag13 — DISMISSED** [review]

File: src/main/java/com/google/sps/data/TSP.java
Comment (albert-g3) [inline src/main/java/com/google/sps/data/TSP.java:19]: outdated comment?

**Review by TylerHernandez — DISMISSED** [review]

File: src/main/java/com/google/sps/data/TSP.java
Comment (mutisog) [inline src/main/java/com/google/sps/data/TSP.java:1]: nit: 2020

File: src/main/java/com/google/sps/data/TSP.java
Comment (mutisog) [inline src/main/java/com/google/sps/data/TSP.java:19]: This doesn't sound accurate. Could you update?

File: src/main/java/com/google/sps/data/TSP.java
Comment (mutisog) [inline src/main/java/com/google/sps/data/TSP.java:22]: Let's remove commented out code if not needed.

File: src/main/java/com/google/sps/data/TSP.java
Comment (mutisog) [inline src/main/java/com/google/sps/data/TSP.java:25]: To improve readability, suggest using a map name that includes the key and value e.g. <key>To<value>Map e.g storeIdToStoreMap.

File: src/main/java/com/google/sps/data/TSP.java
Comment (mutisog) [inline src/main/java/com/google/sps/data/TSP.java:45]: nit: clean up the incorrect comment markers e.g. " * /* "

Comment (mutisog) [inline src/main/java/com/google/sps/data/TSP.java:45]: 1) nit: fix typos: latitudes, longitudes
2) nit: "A breakdown of ..." should be a new sentence, preceded by a period.

File: src/main/java/com/google/sps/data/TSP.java
Comment (mutisog) [inline src/main/java/com/google/sps/data/TSP.java:69]: rename to travelingSalesperson

File: src/main/java/com/google/sps/data/TSP.java
Comment (mutisog) [inline src/main/java/com/google/sps/data/TSP.java:97]: nit: "0 is the user location" --> "visited[0] is the user location"

File: src/main/webapp/store.html
Comment (mutisog) [inline src/main/webapp/store.html:27]: Good name!

File: src/test/java/com/google/sps/data/TSPTest.java
Comment (mutisog) [inline src/test/java/com/google/sps/data/TSPTest.java:35]: Java best practice is to use uppercase names for static and final variables. EPSILON can be made static, IMO.

**Review by mutisog — DISMISSED** [review]
Left some minor comments, but otherwise LGTM, go ahead and submit after other approvals.

File: src/main/java/com/google/sps/data/TSP.java
Comment (albert-g3) [inline src/main/java/com/google/sps/data/TSP.java:20]: TSP is used in this class but class names should be nouns, describing what the purpose of the class is, e.g. DeliveryRouter

File: src/main/java/com/google/sps/data/TSP.java
Comment (albert-g3) [inline src/main/java/com/google/sps/data/TSP.java:50]: `final` to signify this won't change. Rename to `EARTH_RADIUS` and you won't need the comment

File: src/test/java/com/google/sps/data/TSPTest.java
Comment (albert-g3) [inline src/test/java/com/google/sps/data/TSPTest.java:69]: why not calculate this by calling `constructAdjacencyMatrix`?

File: src/test/java/com/google/sps/data/TSPTest.java
Comment (albert-g3) [inline src/test/java/com/google/sps/data/TSPTest.java:95]: you have expected and actual swapped, https://junit.org/junit4/javadoc/4.12/org/junit/Assert.html#assertEquals(double,%20double,%20double)

File: src/test/java/com/google/sps/data/TSPTest.java
Comment (albert-g3) [inline src/test/java/com/google/sps/data/TSPTest.java:34]: can this be declared in the method it's used? you have adjacencyMatrix as instance variable and then declared again in testAdjacencyMatrix

**Review by albert-g3 — DISMISSED** [review]

File: src/test/java/com/google/sps/data/TSPTest.java
Comment (yjzhang111) [reply src/test/java/com/google/sps/data/TSPTest.java:69]: I wanted to test the TSP algorithm only. If I called `constructAdjacencyMatrix` here, wrong result could be caused by not having the correct matrix, but the algorithm itself if correct.

**Comment by mutisog** [comment]
Have we submitted this to the main branch?

File: src/main/java/com/google/sps/data/TSP.java
Comment (yjzhang111) [reply src/main/java/com/google/sps/data/TSP.java:25]: I'm not too sure what I should change it to since the Map is converted from a JSON string and each map contains name, distance to user, storeID, etc.

**Review by albert-g3 — APPROVED** [review]

**Review by dag13 — APPROVED** [review]

**Review by mutisog — APPROVED** [review]

**Review by TylerHernandez — APPROVED** [review]

**Pull request merged**
